components/camel-solr/src/main/java/org/apache/camel/component/solr/SolrProducer.java

```diff
@@ -0,0 +1,302 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.camel.component.solr;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import jakarta.activation.MimetypesFileTypeMap;
+
+import org.apache.camel.Exchange;
+import org.apache.camel.WrappedFile;
+import org.apache.camel.support.DefaultProducer;
+import org.apache.camel.util.ObjectHelper;
+import org.apache.solr.client.solrj.SolrClient;
+import org.apache.solr.client.solrj.SolrQuery;
+import org.apache.solr.client.solrj.SolrServerException;
+import org.apache.solr.client.solrj.request.AbstractUpdateRequest.ACTION;
+import org.apache.solr.client.solrj.request.ContentStreamUpdateRequest;
+import org.apache.solr.client.solrj.request.DirectXmlRequest;
+import org.apache.solr.client.solrj.request.QueryRequest;
+import org.apache.solr.client.solrj.request.UpdateRequest;
+import org.apache.solr.client.solrj.response.QueryResponse;
+import org.apache.solr.common.SolrException;
+import org.apache.solr.common.SolrInputDocument;
+import org.apache.solr.common.params.ModifiableSolrParams;
+
+/**
+ * The Solr producer.
+ */
+public class SolrProducer extends DefaultProducer {
+
+    public SolrProducer(SolrEndpoint endpoint) {
+        super(endpoint);
+    }
+
+    @Override
+    public void process(Exchange exchange) throws Exception {
+
+        String operation = (String) exchange.getIn().getHeader(SolrConstants.OPERATION);
+        if (operation == null) {
+            throw new IllegalArgumentException(SolrConstants.OPERATION + " header is missing");
+        }
+
+        // solr configuration
+        SolrConfiguration solrConfiguration = getEndpoint().getSolrConfiguration(operation);
+
+        // solr client
+        SolrClient solrClient = exchange.getIn().getHeader(SolrConstants.CLIENT, SolrClient.class);
+        if (solrClient == null) {
+            solrClient = getEndpoint().getComponent().getSolrClient(this, solrConfiguration);
+        }
+
+        // solr parameters
+        ModifiableSolrParams solrParams = new ModifiableSolrParams();
+        for (Map.Entry<String, Object> entry : exchange.getIn().getHeaders().entrySet()) {
+            if (entry.getKey().startsWith(SolrConstants.PARAM)) {
+                String paramName = entry.getKey().substring(SolrConstants.PARAM.length());
+                solrParams.add(paramName, entry.getValue().toString());
+            }
+        }
+
+        // solr collection
+        String solrCollection = exchange.getIn().getHeader(SolrConstants.COLLECTION, String.class);
```

Comment:
I'm under the impression this is a mandatory header, no? Probably we should throw `NoSuchHeaderException` if this is missing as well.

Reply:
org.apache.solr.client.solrj.impl.Http2SolrClient.request(SolrRequest<?>, String)
...
    if (ClientUtils.shouldApplyDefaultCollection(collection, solrRequest)) {
      collection = defaultCollection;
    }

There is defaultCollection in SolrClient. So its not mandatory in Exchange headers.